docs/enhancements/multi-rhel-fips.md

```diff
@@ -0,0 +1,137 @@
+run a specified installer binary to generate the manifests required for a
+particular install. These manifests will then be uploaded to whatever storage
+is in use for this deployment (local for on-prem, or s3 for SaaS) and
+assisted-service will take over as usual from there.
+
+Each of these containers should maintain its own cache of installers as
+assisted-service does now and should be built with the required packages to run
+the installer in FIPS compliance mode.
+
+### User Stories
+
+#### Story 1
+
+As a Cluster Creator, I want to install FIPS compliant clusters for any supported OpenShift version
+
+### Implementation Details/Notes/Constraints
+
+#### New Images
+
+Two new container images will need to be built, and published for every release
+footprint we support. These images will be created based on existing
+assisted-service code, but could be split into their own independent projects
+later.
+
+### Risks and Mitigations
+
+Shipping a new image is a non-trivial process. This may take more time than we
+have to set up. We could likely get away with using the existing
+assisted-service image with a different entrypoint for one of the runner images,
+but that still requires us to publish a new one for the architecture
+assisted-service will not be using.
+
+## Design Details [optional]
+
+The new runner containers will expose a HTTP server listening only for local
+connections. assisted-service will POST to this server when it needs manifests
+generated. The runner container will respond with any error that occurred while
+generating the manifests.
+
+### Open Questions
+
+1. What does the API look like for the runner containers? What data should be
+   passed in an API call and what should be configured in the container
+   environment?
+
+### UI Impact
+
+No need to go into great detail about the UI changes that will need to be made.
+However, this is an excellent time to mention 1) if UI changes are required if
+this enhancement were accepted and 2) at a high-level what those UI changes
+would be.
+
+### Test Plan
+
+FIPS and regular installs should be tested for all supported OpenShift versions.
+Since this should be mostly transparent to the user, regression testing in
+addition to testing 4.16 with and without FIPS should be sufficient.
+
+## Drawbacks
+
+This is a complicated change in architecture something simpler might be more
+desirable.
+
+Creating two additional containers in a pod makes the assisted service more
+expensive to scale.
+
+Creating, maintaining, and releasing additional images is a non-trivial amount
+of additional work.
+
+## Alternatives
+
+Hive is investigating using the container image from the release payload to run
+the installer. This seems possible, but doesn't make much sense if we want to
+run a persistent service rather than a job per cluster. Running a job per
+cluster wouldn't scale particularly well and would also be an even larger
+architectural change.
```

Comment:
I don't think we support assisted-installer on podman.
Also at some point podman will get it: https://github.com/containers/podman/issues/17011.

The assisted-service can still be persistent it can use k8s jobs for executing the openshift-installer create-ignition manifests commands.
Why using a job is considered "a larger architectural change" ?

Reply:
We don't offer normal product support for it. But as a project: it works, we invite people to try it, and people use and depend on it.

Reply:
>Why using a job is considered "a larger architectural change" ?

I suppose it's not that much larger really, but it makes the runtime more complicated.
Running jobs over the kube API feels heavier to me than making an http call to a persistently running container in the same pod (effectively an RPC).

Reply:
It would be more difficult to arrange for a shared filesystem with a Job.

You also have to account for all of the things that could happen when scheduling the job, including that it might end up on a node that fails, or on a node that gets drained right as it starts, etc. It would definitely require an asynchronous approach to watching and waiting for the work to be completed.

Reply:
[Here's the cluster-version operator using Jobs to pull down release images][1], and it includes asynchronous watching (which you need for installer runs anyway?) and a shared volume (for getting data back from the Job's Pod to the Job-launching container).  Our volume-sharing is [`hostPath`][2], so we schedule to the same node where the Job-launching container is running, although obviously you could use a cloud-ier volume if you didn't want that requirement.  

[1]: https://github.com/openshift/cluster-version-operator/blob/0b3f507632ce4705702fc725614bd22d25d6686c/pkg/cvo/updatepayload.go#L186-L339
[2]: https://kubernetes.io/docs/concepts/storage/volumes/#hostpath

Reply:
Thanks for the suggestions. We have multiple non-k8s runtime environments that need to keep working, so that's a limiting factor.

We're only using the installer binary to generate artifacts, so we do not need to do asynchronous monitoring of its work or progress. We're not using it to carry out actual provisioning.

Reply:
If we do go with `Jobs` then both ABI and podman would either need to stick with the current method of running the installer on the service container (which might be fine for ABI since it will always be using the matching image, and podman because we'll never support a FIPS cluster that way) or find some other way to spin up the image from the release to run the installer. In any case we end up with needing to support and maintain two methods for running the installer which isn't ideal.

I'm also worried about just how many jobs we might end up running at once.
I don't know exactly what the peak number of installs at a time is on the SaaS, but I know it's at least several hundred simultaneously and several thousand in a few hours when the scale team is testing the on-prem deployment through MCE, @akrzos likely has those numbers.

Reply:
While we don't offer normal product support for the podman runtime method, I don't see a reason why someone couldn't use it on a FIPS-enabled system to install a FIPS-enabled cluster.

Reply:
> While we don't offer normal product support for the podman runtime method, I don't see a reason why someone couldn't use it on a FIPS-enabled system to install a FIPS-enabled cluster.
I don't disagree, but the question is more whether we need the upstream project to support installing both pre-4.16 and post-4.16 clusters with FIPS from the same podman deployment and not ask the user to create separate podman deployments. That sounds like the kind of requirement that products have but which upstream projects can dispense with.

Reply:
For what it's worth I added some more detail about these alternatives to the actual enhancement

Reply:
We don't build a different assisted-installer service to install one version of openshift vs. another. It's just one service that theoretically covers them all.

In theory we could make two parallel builds available and tell users to run one stack for version <=4.15 and the other for versions >4.16, but that seems like more work than just solving the problem properly for everyone whether they're running the upstream or downstream build.